lib/concurrency/Prefix_Refinement.thy

```diff
@@ -938,163 +1189,614 @@ theorem prefix_refinement_bind_general[rule_format]:
+          shouldn't bother with this?\<close>
+lemma prefix_refinement_unless:
+  "\<lbrakk>G = G'; prefix_refinement sr isr isr rvr AR R P Q a c; rvr () ()\<rbrakk>
+   \<Longrightarrow> prefix_refinement sr isr isr rvr AR R (\<lambda>x y. \<not> G \<longrightarrow> P x y) (\<lambda>x y. \<not> G' \<longrightarrow> Q x y)
+         (unless G a) (unless G' c)"
+  by (simp add: unless_def prefix_refinement_when)
+
+\<comment> \<open>FIXME: an equivalent doesn't seem to exist for corres, do we want it to or does it mean we
+          shouldn't bother with this?\<close>
+lemma prefix_refinement_unlessE:
+  "\<lbrakk>G = G'; prefix_refinement sr isr isr (f \<oplus> rvr) AR R P Q a c; rvr () ()\<rbrakk>
+   \<Longrightarrow> prefix_refinement sr isr isr (f \<oplus> rvr) AR R (\<lambda>x y. \<not> G \<longrightarrow> P x y) (\<lambda>x y. \<not> G' \<longrightarrow> Q x y)
+         (unlessE G a) (unlessE G' c)"
+  by (simp add: unlessE_whenE prefix_refinement_whenE)
+
+lemma prefix_refinement_if_r:
+  "\<lbrakk>G' \<Longrightarrow> prefix_refinement sr isr osr rvr AR R P Q a c;
+    \<not>G' \<Longrightarrow> prefix_refinement sr isr osr rvr AR R P Q' a d \<rbrakk>
+   \<Longrightarrow> prefix_refinement sr isr osr rvr AR R
+               P (if G' then Q else Q')
+               a (if G' then c else d)"
+  by simp
+
+lemma prefix_refinement_if3:
+  "\<lbrakk>G = G'; G' \<Longrightarrow> prefix_refinement sr isr osr rvr AR R P Q a c;
+    \<not>G' \<Longrightarrow> prefix_refinement sr isr osr rvr AR R P' Q' b d \<rbrakk>
+   \<Longrightarrow> prefix_refinement sr isr osr rvr AR R
+               (if G then P else P') (if G' then Q else Q')
+               (if G then a else b) (if G' then c else d)"
+  by simp
+
+lemma prefix_refinement_if_strong:
+  "\<lbrakk>\<And>s0 s t0 t. \<lbrakk>sr s0 t0; isr s t; P'' s0 s; Q'' t0 t\<rbrakk> \<Longrightarrow> G = G';
+    \<lbrakk>G; G'\<rbrakk> \<Longrightarrow> prefix_refinement sr isr osr rvr AR R P Q a c;
+    \<lbrakk>\<not>G; \<not>G'\<rbrakk> \<Longrightarrow> prefix_refinement sr isr osr rvr AR R P' Q' b d \<rbrakk>
+   \<Longrightarrow> prefix_refinement sr isr osr rvr AR R
+               (P'' and (if G then P else P')) (Q'' and (if G' then Q else Q'))
+               (if G then a else b) (if G' then c else d)"
+  by (fastforce simp: prefix_refinement_def)
+
+\<comment> \<open>FIXME: corres equivalent isn't used, keep or delete?\<close>
+lemmas prefix_refinement_if_strong' =
+  prefix_refinement_if_strong[where P=P and P'=P and P''=P for P,
+                              where Q=Q and Q'=Q and Q''=Q for Q, simplified]
+
+
+\<comment> \<open>FIXME: Put more thought into whether we want this section, and if not what alternative rules
+          would we want. The comment copied from Corres_UL suggests we might not want them.
+          They would be a fair bit more complicated to prove for prefix_refinement.\<close>
+section \<open>Some equivalences about liftM and other useful simps\<close>
```

Comment:
Good to remove this section -- even if they were easy to prove, it'd be better not to introduce this temptation for prefix refinement.

Reply:
Could we want non-simp rules for use when the return relation is already instantiated?